File: frontend/src/api/mcp-service.ts
import { get, post, put, del } from '@/utils/request'

export interface MCPService {
  id: string
  tenant_id?: number
  name: string
  description: string
  enabled: boolean
  transport_type: 'sse' | 'http-streamable' | 'stdio'
  url?: string // Optional: required for SSE/HTTP Streamable
  headers?: Record<string, string>
  auth_config?: {
    // Authentication strategy. Empty/absent means none. "oauth" enables the
    // per-user OAuth2 authorization-code flow (zero-config: discovery +
    // dynamic client registration).
    auth_type?: '' | 'api_key' | 'bearer' | 'oauth'
    // Secret fields (api_key, token) are NEVER returned by the server in
    // this shape — they live behind the /credentials subresource. The
    // optional-property typing remains so create-mode payloads can still
    // carry them in the initial POST body.
    api_key?: string
    // Header name carrying api_key when auth_type is "api_key". Non-secret;
    // empty defaults to "X-API-Key". Lets services expecting the key in a
    // different header (e.g. raw token in "Authorization") work.
    api_key_header?: string
    token?: string
    custom_headers?: Record<string, string>
    // OAuth-only, non-secret configuration.
    scopes?: string[]
    auth_server_metadata_url?: string
  }
  advanced_config?: {
    timeout?: number
    retry_count?: number
    retry_delay?: number
  }
  stdio_config?: {
    command: 'uvx' | 'npx' // Command: uvx or npx
    args: string[] // Command arguments array
  }
  env_vars?: Record<string, string> // Environment variables for stdio transport
  is_builtin?: boolean // Whether this is a builtin MCP service
  // Per-field "configured?" map embedded on the main response (server-side
  // dto.MCPServiceResponse.Credentials). Drives the CredentialResource card
  // without a follow-up GET. Absent for builtin services.
  credentials?: Record<McpCredentialField, CredentialFieldMetadata>
  created_at?: string
  updated_at?: string
}

export interface MCPTool {
  name: string
  description: string
  inputSchema: Record<string, any>
  require_approval?: boolean
  enabled?: boolean
}

export interface MCPToolApprovalRow {
  id: string
  tenant_id?: number
  service_id: string
  tool_name: string
  require_approval: boolean
  enabled: boolean
}

export interface MCPResource {
  uri: string
  name: string
  description?: string
  mimeType?: string
}

export interface MCPTestResult {
  success: boolean
  message?: string
  description?: string
  // Set when the server requires OAuth (RFC 9728) but the service was not
  // configured for it — the UI guides the user to switch to OAuth 2.0.
  oauth_required?: boolean
  tools?: MCPTool[]
  resources?: MCPResource[]
}

// List all MCP services
export async function listMCPServices(): Promise<MCPService[]> {
  const response: any = await get('/api/v1/mcp-services')
  return response.data || []
}

// Get a single MCP service by ID
export async function getMCPService(id: string): Promise<MCPService> {
  const response: any = await get(`/api/v1/mcp-services/${id}`)
  return response.data
}

// Create a new MCP service
export async function createMCPService(data: Partial<MCPService>): Promise<MCPService> {
  const response: any = await post('/api/v1/mcp-services', data)
  return response.data
}

// Update an existing MCP service
export async function updateMCPService(id: string, data: Partial<MCPService>): Promise<MCPService> {
  const response: any = await put(`/api/v1/mcp-services/${id}`, data)
  return response.data
}

// Delete an MCP service
export async function deleteMCPService(id: string): Promise<void> {
  await del(`/api/v1/mcp-services/${id}`)
}

// Test MCP service connection
export async function testMCPService(id: string): Promise<MCPTestResult> {
  const response: any = await post(`/api/v1/mcp-services/${id}/test`, {})
  // 后端返回格式: { success: true, data: MCPTestResult }
  // response interceptor 已经返回了 data，所以 response 就是 { success: true, data: {...} }
  if (response && response.data) {
    return response.data
  }
  // 如果格式不对，尝试直接返回 response（可能是直接返回的数据）
  return response
}

// Get tools from an MCP service
export async function getMCPServiceTools(id: string): Promise<MCPTool[]> {
  const response: any = await get(`/api/v1/mcp-services/${id}/tools`)
  return response.data || []
}

// Get resources from an MCP service
export async function getMCPServiceResources(id: string): Promise<MCPResource[]> {
  const response: any = await get(`/api/v1/mcp-services/${id}/resources`)
  return response.data || []
}

/** Persisted per-tool human-approval flags (issue #1173) */
export async function getMCPToolApprovals(serviceId: string): Promise<MCPToolApprovalRow[]> {
  const response: any = await get(`/api/v1/mcp-services/${serviceId}/tool-approvals`)
  return response.data || []
}

export async function setMCPToolApproval(serviceId: string, toolName: string, requireApproval: boolean): Promise<void> {
  await put(`/api/v1/mcp-services/${serviceId}/tool-approvals/${encodeURIComponent(toolName)}`, {
    require_approval: requireApproval
  })
}

export async function setMCPToolEnabled(serviceId: string, toolName: string, enabled: boolean): Promise<void> {
  await put(`/api/v1/mcp-services/${serviceId}/tool-approvals/${encodeURIComponent(toolName)}`, { enabled })
}

// ----------------------------------------------------------------------------
// Credential subresource (issue #988 follow-up).
//
// Secrets travel through a dedicated /credentials endpoint instead of the
// main MCP PUT body. "Is this configured?" metadata is embedded on the main
// MCPService response (MCPService.credentials), so there is no GET on this
// endpoint — only PUT (write) and DELETE (clear). Both trigger an MCP
// client reconnect server-side.
// ----------------------------------------------------------------------------

export type McpCredentialField = 'api_key' | 'token'

export interface CredentialFieldMetadata {
  configured: boolean
}

export interface McpCredentialsResponse {
  fields: Record<McpCredentialField, CredentialFieldMetadata>
}

export async function putMCPCredentials(
  serviceId: string,
  body: Partial<Record<McpCredentialField, string>>
): Promise<McpCredentialsResponse> {
  const response: any = await put(`/api/v1/mcp-services/${serviceId}/credentials`, body)
  return (response.data ?? response) as McpCredentialsResponse
}

export async function deleteMCPCredentialField(
  serviceId: string,
  field: McpCredentialField
): Promise<void> {
  await del(`/api/v1/mcp-services/${serviceId}/credentials/${field}`)
}

// ----------------------------------------------------------------------------
// Per-user OAuth2 authorization-code flow.
//
// The user authorizes a service once; the backend stores their access/refresh
// token (per tenant + user + service) and refreshes it transparently. The
// callback is a public backend route that the third-party authorization
// server redirects to.
// ----------------------------------------------------------------------------

// Path of the public backend OAuth callback (registered outside /mcp-services
// to avoid a route conflict, and allow-listed for no-auth in the backend).
export const MCP_OAUTH_CALLBACK_PATH = '/api/v1/mcp-oauth/callback'

export interface MCPOAuthAuthorization {
  authorizationUrl: string
  authorizationAttempt: string
}

export type MCPOAuthTokenState = 'authorized' | 'refreshable' | 'reauth_required' | 'pending'

export interface MCPOAuthStatus {
  authorized: boolean
  state: MCPOAuthTokenState
  refresh_available: boolean
  expires_at?: string
}

// Begin authorization for the current user. The attempt id binds polling to
// this popup, so an older stored token cannot be mistaken for fresh consent.
export async function getMCPOAuthAuthorizeURL(
  serviceId: string,
  body: { redirect_uri: string; frontend_redirect?: string }
): Promise<MCPOAuthAuthorization> {
  const response: any = await post(`/api/v1/mcp-services/${serviceId}/oauth/authorize-url`, body)
  const data = response.data ?? response
  return {
    authorizationUrl: data?.authorization_url ?? '',
    authorizationAttempt: data?.authorization_attempt ?? '',
  }
}

// Whether the current user has authorized this service.
export async function getMCPOAuthStatus(
  serviceId: string,
  authorizationAttempt?: string,
): Promise<boolean> {
  const query = authorizationAttempt
    ? `?authorization_attempt=${encodeURIComponent(authorizationAttempt)}`
    : ''
  const response: any = await get(`/api/v1/mcp-services/${serviceId}/oauth/status${query}`)
  return Boolean((response.data ?? response)?.authorized)
}

// Full lifecycle status for management surfaces. Expired access tokens with a
// refresh token are "refreshable", not falsely presented as already usable.
export async function getMCPOAuthAuthorizationStatus(serviceId: string): Promise<MCPOAuthStatus> {
  const response: any = await get(`/api/v1/mcp-services/${serviceId}/oauth/status`)
  const data = response.data ?? response
  return {
    authorized: Boolean(data?.authorized),
    state: data?.state ?? 'reauth_required',
    refresh_available: Boolean(data?.refresh_available),
    expires_at: data?.expires_at,
  }
}

// Revoke the current user's token (forces re-authorization).
export async function revokeMCPOAuthToken(serviceId: string): Promise<void> {
  await del(`/api/v1/mcp-services/${serviceId}/oauth/token`)
}

export async function resolveToolApproval(
  pendingId: string,
  body: { decision: 'approve' | 'reject'; modified_args?: Record<string, unknown>; reason?: string }
): Promise<void> {
  await post(`/api/v1/agent/tool-approvals/${encodeURIComponent(pendingId)}`, body)
}

// Resume an agent run that paused on an in-conversation MCP OAuth prompt.
// Call after the per-user authorization popup completes; the backend verifies
// the token exists before unblocking the paused tool call.
export async function resolveMCPOAuth(
  pendingId: string,
  body: { service_id: string; decision?: 'authorize' | 'cancel' }
): Promise<void> {
  await post(`/api/v1/agent/mcp-oauth-resolutions/${encodeURIComponent(pendingId)}`, body)
}

export async function cancelMCPOAuth(pendingId: string): Promise<void> {
  await post(`/api/v1/agent/mcp-oauth-resolutions/${encodeURIComponent(pendingId)}/cancel`, {})
}

File: client/mcp_service.go
package client

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// MCPTransportType represents the transport type for MCP service
type MCPTransportType string

const (
	MCPTransportSSE            MCPTransportType = "sse"
	MCPTransportHTTPStreamable MCPTransportType = "http-streamable"
	MCPTransportStdio          MCPTransportType = "stdio"
)

// MCPService represents an MCP service configuration
type MCPService struct {
	ID                string             `json:"id"`
	TenantID          uint64             `json:"tenant_id"`
	Name              string             `json:"name"`
	Description       string             `json:"description"`
	UsageInstructions string             `json:"usage_instructions,omitempty"`
	Enabled           bool               `json:"enabled"`
	TransportType     MCPTransportType   `json:"transport_type"`
	URL               *string            `json:"url,omitempty"`
	Headers           map[string]string  `json:"headers"`
	AuthConfig        *MCPAuthConfig     `json:"auth_config"`
	AdvancedConfig    *MCPAdvancedConfig `json:"advanced_config"`
	StdioConfig       *MCPStdioConfig    `json:"stdio_config,omitempty"`
	EnvVars           map[string]string  `json:"env_vars,omitempty"`
	IsBuiltin         bool               `json:"is_builtin"`
	CreatedAt         string             `json:"created_at"`
	UpdatedAt         string             `json:"updated_at"`
	Catalog           *MCPCatalogSummary `json:"catalog,omitempty"`
}

// MCPCatalogSummary is the list-card view of a saved MCP directory.
type MCPCatalogSummary struct {
	ToolCount int    `json:"tool_count"`
	Stale     bool   `json:"stale"`
	SyncedAt  string `json:"synced_at"`
}

// MCPAuthConfig represents authentication configuration for MCP service.
//
// Secret fields (APIKey, Token) are accepted on create but are never returned
// by the server. To mutate credentials on an existing service, use the
// dedicated /credentials subresource — see the MCP credentials API for the
// PUT / DELETE shape. Sending secret fields in a main PUT body is silently
// ignored server-side.
type MCPAuthConfig struct {
	APIKey        string            `json:"api_key,omitempty"`
	Token         string            `json:"token,omitempty"`
	CustomHeaders map[string]string `json:"custom_headers,omitempty"`
}

// MCPAdvancedConfig represents advanced configuration for MCP service
type MCPAdvancedConfig struct {
	Timeout    int `json:"timeout"`
	RetryCount int `json:"retry_count"`
	RetryDelay int `json:"retry_delay"`
}

// MCPStdioConfig represents stdio transport configuration
type MCPStdioConfig struct {
	Command string   `json:"command"`
	Args    []string `json:"args"`
}

// MCPTool represents a tool exposed by an MCP service
type MCPTool struct {
	Name        string          `json:"name"`
	Description string          `json:"description"`
	InputSchema json.RawMessage `json:"inputSchema"`
}

// MCPResource represents a resource exposed by an MCP service
type MCPResource struct {
	URI         string `json:"uri"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	MimeType    string `json:"mimeType,omitempty"`
}

// MCPTestResult represents the result of testing an MCP service connection
type MCPTestResult struct {
	Success     bool           `json:"success"`
	Message     string         `json:"message,omitempty"`
	Description string         `json:"description,omitempty"`
	Tools       []*MCPTool     `json:"tools,omitempty"`
	Resources   []*MCPResource `json:"resources,omitempty"`
}

// CreateMCPService creates a new MCP service
func (c *Client) CreateMCPService(ctx context.Context, service *MCPService) (*MCPService, error) {
	resp, err := c.doRequest(ctx, http.MethodPost, "/api/v1/mcp-services", service, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Success bool        `json:"success"`
		Data    *MCPService `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// ListMCPServices lists all MCP services for the current tenant
func (c *Client) ListMCPServices(ctx context.Context) ([]*MCPService, error) {
	resp, err := c.doRequest(ctx, http.MethodGet, "/api/v1/mcp-services", nil, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Success bool          `json:"success"`
		Data    []*MCPService `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// GetMCPService gets an MCP service by ID
func (c *Client) GetMCPService(ctx context.Context, serviceID string) (*MCPService, error) {
	resp, err := c.doRequest(ctx, http.MethodGet, fmt.Sprintf("/api/v1/mcp-services/%s", serviceID), nil, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Success bool        `json:"success"`
		Data    *MCPService `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// UpdateMCPService updates an MCP service
func (c *Client) UpdateMCPService(ctx context.Context, serviceID string, updates map[string]interface{}) (*MCPService, error) {
	resp, err := c.doRequest(ctx, http.MethodPut, fmt.Sprintf("/api/v1/mcp-services/%s", serviceID), updates, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Success bool        `json:"success"`
		Data    *MCPService `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// DeleteMCPService deletes an MCP service
func (c *Client) DeleteMCPService(ctx context.Context, serviceID string) error {
	resp, err := c.doRequest(ctx, http.MethodDelete, fmt.Sprintf("/api/v1/mcp-services/%s", serviceID), nil, nil)
	if err != nil {
		return err
	}
	return parseResponse(resp, nil)
}

// TestMCPService tests an MCP service connection
func (c *Client) TestMCPService(ctx context.Context, serviceID string) (*MCPTestResult, error) {
	resp, err := c.doRequest(ctx, http.MethodPost, fmt.Sprintf("/api/v1/mcp-services/%s/test", serviceID), nil, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Success bool           `json:"success"`
		Data    *MCPTestResult `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// GetMCPServiceTools gets the tools provided by an MCP service
func (c *Client) GetMCPServiceTools(ctx context.Context, serviceID string) ([]*MCPTool, error) {
	resp, err := c.doRequest(ctx, http.MethodGet, fmt.Sprintf("/api/v1/mcp-services/%s/tools", serviceID), nil, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Success bool       `json:"success"`
		Data    []*MCPTool `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// MCPMetadata is a persisted MCP tool directory. A nil result from GetMCPMetadata
// means the service has never been synchronized. Stale is true when the saved
// connection no longer matches the current configuration.
type MCPMetadata struct {
	ServiceID         string     `json:"service_id"`
	Tools             []*MCPTool `json:"tools"`
	Instructions      string     `json:"instructions"`
	ServerName        string     `json:"server_name"`
	ServerVersion     string     `json:"server_version"`
	ServerDescription string     `json:"server_description"`
	SyncedAt          string     `json:"synced_at"`
	Stale             bool       `json:"stale"`
}

// GetMCPMetadata reads the saved tool directory without connecting upstream.
// Server route: GET /api/v1/mcp-services/{id}/metadata.
func (c *Client) GetMCPMetadata(ctx context.Context, serviceID string) (*MCPMetadata, error) {
	resp, err := c.doRequest(ctx, http.MethodGet, fmt.Sprintf("/api/v1/mcp-services/%s/metadata", serviceID), nil, nil)
	if err != nil {
		return nil, err
	}
	var result struct {
		Success bool         `json:"success"`
		Data    *MCPMetadata `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// RefreshMCPMetadata connects upstream and atomically replaces the saved directory.
// OAuth services store a snapshot for the calling user (Viewer+). Static-auth
// services write a tenant-wide snapshot and require Admin, or an API key that
// can manage MCP services.
// Server route: POST /api/v1/mcp-services/{id}/metadata/refresh.
func (c *Client) RefreshMCPMetadata(ctx context.Context, serviceID string) (*MCPMetadata, error) {
	path := fmt.Sprintf("/api/v1/mcp-services/%s/metadata/refresh", serviceID)
	resp, err := c.doRequest(ctx, http.MethodPost, path, nil, nil)
	if err != nil {
		return nil, err
	}
	var result struct {
		Success bool         `json:"success"`
		Data    *MCPMetadata `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// GetMCPServiceResources gets the resources provided by an MCP service
func (c *Client) GetMCPServiceResources(ctx context.Context, serviceID string) ([]*MCPResource, error) {
	resp, err := c.doRequest(ctx, http.MethodGet, fmt.Sprintf("/api/v1/mcp-services/%s/resources", serviceID), nil, nil)
	if err != nil {
		return nil, err
	}

	var result struct {
		Success bool           `json:"success"`
		Data    []*MCPResource `json:"data"`
	}
	if err := parseResponse(resp, &result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

// ResolveToolApprovalRequest is the body for resolving a pending tool-approval
// raised during an agent run (session ask). Decision is "approve" or "reject".
// ModifiedArgs optionally replaces the tool call arguments on approve; it must
// be a JSON object when present.
type ResolveToolApprovalRequest struct {
	Decision     string          `json:"decision"`
	Reason       string          `json:"reason,omitempty"`
	ModifiedArgs json.RawMessage `json:"modified_args,omitempty"`
}

// ResolveToolApproval resolves a pending tool approval by id.
// Server route: POST /api/v1/agent/tool-approvals/{pending_id}.
func (c *Client) ResolveToolApproval(ctx context.Context, pendingID string, req *ResolveToolApprovalRequest) error {
	path := fmt.Sprintf("/api/v1/agent/tool-approvals/%s", pendingID)
	resp, err := c.doRequest(ctx, http.MethodPost, path, req, nil)
	if err != nil {
		return err
	}
	var response struct {
		Success bool   `json:"success"`
		Message string `json:"message,omitempty"`
	}
	return parseResponse(resp, &response)
}
